_posts/2019-04-03-psql-performance-and-diagnostic-tricks.md

```diff
@@ -0,0 +1,258 @@
+---
+layout: post
+title:  "PostgreSQL Performance and Diagnostic Tricks"
+date:   2019-04-03
+categories: database performance
+author: Nathan Lilienthal
+---
+
+Here at Forward Financing we use [PostgreSQL](postgresql) heavily for our main
+application databases. It's a modern, feature rich, and performant database.
+But it's not always performant, and if you're not watching, it can come back to
+bite you.
+
+Starting the week of March 25, 2019, we noticed elevated timeout rates on one
+of our main applications. This was quickly causing a bit of talk around office.
+
+In this post we'll go over a few of the steps, related to our database, that we
+took to resolve our issue. This application's DB has a number of good easy low
+hanging fruit to pick off. These techniques aren't overly advanced, yet
+resulted in a dramatic reduction in some queries, and ultimately response times
+/ timeouts.
+
+> 1. [Metrics](#metrics)
+>     1. [Scout Metrics](#scout-metrics)
+>     2. [Heroku Database Metrics](#heroku-database-metrics)
+>     3. [`psql` Metrics](#psql-metrics)
+> 2. [Improving Query Performance](#improving-query-performance)
+>     1. [Missing indices](#missing-indices)
+>     2. [N+1 Query](#remove-n1-query)
+> 3. [Improving Database Health](#improving-database-health)
+
+# Metrics
+
+This is mostly a post about DB monitoring, diagnostics, and optimization.
+However, this is the result of an investigation into a performance issue we
+were facing with an application, which was causing both H12 (timeout) errors in
+Heroku, but also just very poor response times.
+
+- Server Respose Time
+- Query Execution Time / Count
+- Cache Hit Rates
+- Table Bloat
+
+## Scout Metrics
+
+First, before we had identified any potential fixes, and we were just hearing
+reports of pages timing out a lot more than normal, we could look at
+[Scout][scout].
+
+The image below is kinda a sneak peek at the performance gains we will achieve
+throughout this post (so far).
+
+![]({{"/assets/2019-04-03-psql-performance-and-diagnostic-tricks/compare_admin_applications_show.png" | absolute_url}})
+
+Here, the solid brown "total" is the optimized average response times, while
+the dotted brown is the "total" for the previous slow code. It shows a -55%
+change in mean response time!
+
+## Heroku Database Metrics
+
+- Most time consuming
+- Slowest execution
+
+The most time consuming queries are the first we started looking at to be
```

Comment:
This paragraph is a little awkward, you switch tenses a few times. Does this sound any better?

```
We started optimizing the most time consuming queries, since the goal
was to reduce the number of users impacted by timeouts. Speeding up the
slowest queries puts less pressure on the database allowing other queries to
speed up naturally.
```

Reply:
Yea I agree, this paragraph is a massive run on.

I've tweaked your suggestion slightly.